packages/angular/ssr/src/app.ts

```diff
@@ -73,7 +121,60 @@ export class AngularServerApp {
       return Response.redirect(new URL(redirectTo, url), 302);
     }
 
-    return render(this, request, serverContext, requestContext);
+    const isSsrMode = serverContext === ServerRenderContext.SSR;
+    const responseInit: ResponseInit = {};
+    const platformProviders: StaticProvider = [
+      {
+        provide: SERVER_CONTEXT,
+        useValue: serverContext,
+      },
+    ];
+
+    if (isSsrMode) {
+      platformProviders.push(
+        {
+          provide: REQUEST,
+          useValue: request,
+        },
+        {
+          provide: REQUEST_CONTEXT,
+          useValue: requestContext,
+        },
+        {
+          provide: RESPONSE_INIT,
+          useValue: responseInit,
+        },
+      );
+    }
+
+    if (typeof ngDevMode === 'undefined' || ngDevMode) {
+      // Need to clean up GENERATED_COMP_IDS map in `@angular/core`.
+      // Otherwise an incorrect component ID generation collision detected warning will be displayed in development.
+      // See: https://github.com/angular/angular-cli/issues/25924
+      ɵresetCompiledComponents();
+    }
+
+    // An Angular Console Provider that does not print a set of predefined logs.
+    platformProviders.push({
+      provide: ɵConsole,
+      // Using `useClass` would necessitate decorating `Console` with `@Injectable`,
+      // which would require switching from `ts_library` to `ng_module`. This change
+      // would also necessitate various patches of `@angular/bazel` to support ESM.
+      useFactory: () => new Console(),
+    });
+
+    const { manifest, hooks, assets } = this;
+
+    let html = await assets.getIndexServerHtml();
+    // Skip extra microtask if there are no pre hooks.
+    if (hooks.has('html:transform:pre')) {
+      html = await hooks.run('html:transform:pre', { html });
+    }
+
+    return new Response(
+      await renderAngular(html, manifest.bootstrap(), new URL(request.url), platformProviders),
```

Comment:
Question: Can we pass through `request.signal` anywhere in here to abort these operations? Otherwise we're somewhat wasting effort doing rendering work for a request which has already been canceled.

* I think `getIndexServerHtml` is ultimately a dynamic import, so we probably can't abort that.
* We could pass it through to the hook callbacks if we want to add it to the public API there.
* `renderAngular` is probably the most important operation to abort. Can we maybe just stop waiting for stability?

If it's tricky to support this, an easy workaround might just be `if (request.signal.aborted) throw new DOMException('AbortError');` to avoid starting rendering at all. Not as comprehensive, but it's at least something.

Reply:
Let talk about this during our sync.